Placement new warnings generated under GCC 11.1

### Godot version

master `6e87d62873`

### System information

Manjaro Linux (kernel 5.13), GCC 11.1

### Issue description

The following warning is generated almost all the time during the build:
```
./core/templates/cowdata.h: In instantiation of 'uint32_t CowData<T>::_copy_on_write() [with T = Face3; uint32_t = unsigned int]':
./core/templates/cowdata.h:276:16:   required from 'Error CowData<T>::resize(int) [with T = Face3]'
./core/templates/vector.h:88:51:   required from 'Error Vector<T>::resize(int) [with T = Face3]'
./core/variant/method_ptrcall.h:410:13:   required from here
./core/templates/cowdata.h:235:30: warning: placement new constructing an object of type 'SafeNumeric<unsigned int>' and size '4' in a region of type 'uint32_t*' {aka 'unsigned int*'} and size '0' [-Wplacement-new=]
  235 |                 new (mem_new - 2) SafeNumeric<uint32_t>(1); //refcount
      |                      ~~~~~~~~^~~
./core/templates/cowdata.h:233:27: note: at offset -8 from 'mem_new' declared here
  233 |                 uint32_t *mem_new = (uint32_t *)Memory::alloc_static(_get_alloc_size(current_size), true);
      |                           ^~~~~~~
```

Not entirely sure it matters, as it is just a warning, but I find this part here suspicious:
>in a region of type 'uint32_t*' {aka 'unsigned int*'} and size '0'

I have bisected this to commit 98ceb60eb4. Will attempt a fix, but I am not entirely sure what the intent of the code in that place is, so I might ask a few things if I am stuck.

### Steps to reproduce

Under latest Manjaro with GCC 11.1, run this in the project root:

`scons platform=linuxbsd target=release_debug use_lto=yes -j4`


### Minimal reproduction project

Not applicable here.

--- a/core/templates/cowdata.h
+++ b/core/templates/cowdata.h
@@ -49,6 +49,12 @@ class VMap;
 SAFE_NUMERIC_TYPE_PUN_GUARANTEES(uint32_t)
 #endif
 
+// Silence a false positive warning (see GH-52119).
+#if defined(__GNUC__) && !defined(__clang__)
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wplacement-new"
+#endif
+
 template <class T>
 class CowData {
 	template <class TV>
@@ -379,5 +385,9 @@ template <class T>
 CowData<T>::~CowData() {
 	_unref(_ptr);
 }
+
+#if defined(__GNUC__) && !defined(__clang__)
+#pragma GCC diagnostic pop
+#endif
 
 #endif // COWDATA_H

